src/backend/base/langflow/services/database/models/transactions/crud.py

```diff
@@ -26,14 +27,51 @@ async def get_transactions_by_flow_id(
 
 
 async def log_transaction(db: AsyncSession, transaction: TransactionBase) -> TransactionTable:
+    """Log a transaction and maintain a maximum number of transactions in the database.
+
+    This function logs a new transaction into the database and ensures that the number of transactions
+    does not exceed the maximum limit specified in the settings. If the number of transactions exceeds
+    the limit, the oldest transactions are deleted to maintain the limit.
+
+    Args:
+        db: Database session
+        transaction: Transaction data to log
+
+    Returns:
+        The created TransactionTable entry
+
+    Raises:
+        IntegrityError: If there is a database integrity error
+    """
     table = TransactionTable(**transaction.model_dump())
-    db.add(table)
+
     try:
+        # Get max entries setting
+        max_entries = get_settings_service().settings.max_transactions_to_keep
+
+        # Delete older entries in a single transaction
+        delete_older = delete(TransactionTable).where(
+            TransactionTable.flow_id == transaction.flow_id,
+            TransactionTable.id.in_(
+                select(TransactionTable.id)
+                .where(TransactionTable.flow_id == transaction.flow_id)
+                .order_by(TransactionTable.timestamp.desc())
+                .offset(max_entries - 1)  # Keep newest max_entries-1 plus the one we're adding
```

Comment:
What happens if there are no entries from the `select` and the `_in()` is empty ?

Reply:
It won't delete anything.